### PR TITLE
 Bug 28957665: Fixed psObject List Enumeration

### LINKED ISSUE #25321: Update-AzMaintenanceConfiguration always fail with Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration'
### Description

Hi,
I cannot find any way of getting Update-AzMaintenanceConfiguration to work. I am using Az 12.0.0.

```
Get-InstalledModule -Name Az -AllVersions | select Name,Version

Name Version
---- -------
Az   12.0.0

```
```powershell
$MaintenanceConfig = Get-AzMaintenanceConfiguration -name 'aum-pelo01'
Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
```
Here I get an existing configuration and send it to Update-AzMaintenanceConfiguration.

It always fail with:
`Update-AzMaintenanceConfiguration: Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' to the type 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' required by parameter 'Configuration'. Specified method is not supported.

The same happens if I make a small change to a parameter:
```
$MaintenanceConfig = Get-AzMaintenanceConfiguration -name 'aum-pelo01'
$MaintenanceConfig[0].WindowParameterKbNumberToInclude += '123456'
Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
```

### Issue script & Debug output

```PowerShell
PS C:\Users\pelo> $DebugPreference='Continue'
PS C:\Users\pelo> Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
Update-AzMaintenanceConfiguration: Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' to the type 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' required by parameter 'Configuration'. Specified method is not supported.
```

### Environment data

```PowerShell
PS C:\Users\pelo> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.2
PSEdition                      Core
GitCommitId                    7.4.2
OS                             Microsoft Windows 10.0.22631
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Module versions

```PowerShell
PS C:\Users\pelo> Get-Module Az*

ModuleType Version    PreRelease Name                                ExportedCommands
---------- -------    ---------- ----                                ----------------
Script     3.0.0                 Az.Accounts                         {Add-AzEnvironment, Clear-AzConfig, Clear-AzContext, Clear-AzDefault…}
Script     1.4.1                 Az.Maintenance                      {Get-AzApplyUpdate, Get-AzConfigurationAssignment, Get-AzMaintenanceConfiguration, Get-Az…
```

### Error output

```PowerShell
PS C:\Users\pelo> $MaintenanceConfig = Get-AzMaintenanceConfiguration -name 'aum-pelo01'
PS C:\Users\pelo> $MaintenanceConfig[0].WindowParameterKbNumberToInclude += '123456'
PS C:\Users\pelo> Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
Update-AzMaintenanceConfiguration: Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' to the type 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' required by parameter 'Configuration'. Specified method is not supported.
PS C:\Users\pelo> Resolve-AzError

   HistoryId: 3

Message        : Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' to the type 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' required by
                 parameter 'Configuration'. Specified method is not supported.
StackTrace     :    at System.Management.Automation.ParameterBinderBase.CoerceTypeAsNeeded(CommandParameterInternal argument, String parameterName, Type toType, ParameterCollectionTypeInformation collec
                 tionTypeInfo, Object currentValue)
                    at System.Management.Automation.ParameterBinderBase.BindParameter(CommandParameterInternal parameter, CompiledCommandParameter parameterMetadata, ParameterBindingFlags flags)
                    at System.Management.Automation.CmdletParameterBinderController.BindParameter(CommandParameterInternal argument, MergedCompiledCommandParameter parameter, ParameterBindingFlags flags
                 )
                    at System.Management.Automation.CmdletParameterBinderController.BindParameter(UInt32 parameterSets, CommandParameterInternal argument, MergedCompiledCommandParameter parameter, Param
                 eterBindingFlags flags)
                    at System.Management.Automation.CmdletParameterBinderController.BindNamedParameter(UInt32 parameterSets, CommandParameterInternal argument, MergedCompiledCommandParameter parameter)
                    at System.Management.Automation.ParameterBinderController.BindNamedParameters(UInt32 parameterSets, Collection`1 arguments)
                    at System.Management.Automation.CmdletParameterBinderController.BindCommandLineParametersNoValidation(Collection`1 arguments)
                    at System.Management.Automation.CmdletParameterBinderController.BindCommandLineParameters(Collection`1 arguments)
                    at System.Management.Automation.CommandProcessor.BindCommandLineParameters()
                    at System.Management.Automation.CommandProcessor.Prepare(IDictionary psDefaultParameterValues)
                    at System.Management.Automation.CommandProcessorBase.DoPrepare(IDictionary psDefaultParameterValues)
                    at System.Management.Automation.Internal.PipelineProcessor.Start(Boolean incomingStream)
                    at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
                 --- End of stack trace from previous location ---
                    at System.Management.Automation.Internal.PipelineProcessor.SynchronousExecuteEnumerate(Object input)
                    at System.Management.Automation.PipelineOps.InvokePipeline(Object input, Boolean ignoreInput, CommandParameterInternal[][] pipeElements, CommandBaseAst[] pipeElementAsts, CommandRedi
                 rection[][] commandRedirections, FunctionContext funcContext)
                    at System.Management.Automation.Interpreter.ActionCallInstruction`6.Run(InterpretedFrame frame)
                    at System.Management.Automation.Interpreter.EnterTryCatchFinallyInstruction.Run(InterpretedFrame frame)
Exception      : System.Management.Automation.ParameterBindingException
InvocationInfo : {Update-AzMaintenanceConfiguration}
Line           : Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
Position       : At line:1 char:100
                 + … Name 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
                 +                                                      ~~~~~~~~~~~~~~~~~~
HistoryId      : 3

Message        : Specified method is not supported.
StackTrace     :    at System.Management.Automation.ParameterBinderBase.CoerceTypeAsNeeded(CommandParameterInternal argument, String parameterName, Type toType, ParameterCollectionTypeInformation collec
                 tionTypeInfo, Object currentValue)
Exception      : System.NotSupportedException
InvocationInfo : {Update-AzMaintenanceConfiguration}
Line           : Update-AzMaintenanceConfiguration -ResourceGroupName 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
Position       : At line:1 char:100
                 + … Name 'pelo-test' -Name 'aum-pelo01' -Configuration $MaintenanceConfig
                 +                                                      ~~~~~~~~~~~~~~~~~~
HistoryId      : 3
```

### PR DESCRIPTION
**Pull request by notyashhh**
## Problem Description:
- The Update-AzMaintenanceConfiguration command was failing with the error:
Cannot convert 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' to the type 'Microsoft.Azure.Commands.Maintenance.Models.PSMaintenanceConfiguration' required by parameter 'Configuration'. Specified method is not supported.

This issue occurred when Get-AzMaintenanceConfiguration was called without the ResourceGroupName parameter, causing the returned object to be of type List<PSMaintenanceConfiguration>, instead of a single PSMaintenanceConfiguration.

## Fix Implemented:
- Modified the Update-AzMaintenanceConfiguration command to handle both cases where Get-AzMaintenanceConfiguration returns a list of configurations or a single configuration object.
- Ensured that the command properly handles and converts the list of configurations into individual PSMaintenanceConfiguration objects when required.

## Verification:
- Verified that Update-AzMaintenanceConfiguration works correctly with both single and multiple maintenance configurations.
Added test cases to cover both scenarios ensuring the command operates as expected.

## Issue Addressed:
- Closes #25321 

## Testing:

- Ran existing tests to confirm no regressions.
- Added new test cases for scenarios involving single and multiple maintenance configurations.

## Impact:
- This fix ensures that users can now successfully update maintenance configurations regardless of whether they specify the ResourceGroupName parameter when calling Get-AzMaintenanceConfiguration.

## Mandatory Checklist

- Please choose the target release of Azure PowerShell. (⚠️**Target release** is a different concept from **API readiness**. Please click below links for details.)
  - [x] [General release](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Public preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Private preview](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] [Engineering build](../blob/main/CONTRIBUTING.md#target-release-types)
  - [ ] No need for a release

- [x] Check this box to confirm: **I have read the [_Submitting Changes_](../blob/main/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](../blob/main/CONTRIBUTING.md) and reviewed the following information:**

* **SHOULD** update `ChangeLog.md` file(s) appropriately
    * For SDK-based development mode, update `src/{{SERVICE}}/{{SERVICE}}/ChangeLog.md`.
        * A snippet outlining the change(s) made in the PR should be written under the `## Upcoming Release` header in the past tense. 
    * For autorest-based development mode, include the changelog in the PR description.
    * Should **not** change `ChangeLog.md` if no new release is required, such as fixing test case only.
* **SHOULD** regenerate markdown help files if there is cmdlet API change. [Instruction](../blob/main/documentation/development-docs/help-generation.md#updating-all-markdown-files-in-a-module)
* **SHOULD** have proper test coverage for changes in pull request.
* **SHOULD NOT** adjust version of module manually in pull request

### CODE REVIEW

**Comment by BethanyZhou** [comment]
Hi @notyashhh, 
- Could you link this PR with the original GitHub issue so that the issue will be closed automatically once the PR gets merge? BTW, ADO item, which is an internal way to track issue, should not be exposed in our public GitHub repo.
- Please check the mandatory checklist in the PR board.

File: src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1 [hunk 1]
```diff
@@ -12,6 +12,84 @@
...
+    $expirationDateTime = "9999-12-31 23:59"
+    
+    try
+    {
+        # Create Resource Group
+        New-AzResourceGroup -Name $resourceGroupName -Location $location
+        Write-Host "Created Resource Group $resourceGroupName in $location"
+
+        # Create Maintenance Configuration
+        $maintenanceConfiguration = New-AzMaintenanceConfiguration -ResourceGroupName $resourceGroupName -Name $maintenanceConfigurationName -MaintenanceScope $maintenanceScope -Location $location -Visibility $visibility -StartDateTime $startDateTime -Timezone $timezone -RecurEvery $recurEvery -Duration $duration -ExpirationDateTime $expirationDateTime
```
Comment (BethanyZhou) [inline src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1:39]: The PR targets to fix issue that Update-AzMaintenanceConfiguration can't receive input from pipeline way if Get-AzMaintenanceConfiguration returns multiple maintenance configurations as list. Please include this scenario into tests.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1 (hunk 1 above)
Comment (notyashhh) [reply src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1:39]: Thanks for pointing this out, I have added this scenario :)

File: src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1 [hunk 2]
```diff
@@ -12,6 +12,93 @@
...
+        # Validate that the returned object is a list
+        if ($maintenanceConfigurations -isnot [System.Collections.IList]) {
+            Write-Host "The retrieved maintenance configuration without ResourceGroupName is of type list."
+        } else {
+            throw "The retrieved maintenance configuration without ResourceGroupName is not of type list."
+        }
+
+        # As a final check, also verify that Update-Configuration can accept the inpu
+        try{
+            Update-AzMaintenanceConfiguration -ResourceGroupName $resourceGroupName -Name $maintenanceConfigurationName -Configuration $retrievedMaintenanceConfiguration
```
Comment (BethanyZhou) [inline src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1:85]: Could you try to pass values via pipeline in test cases? like $retrievedMaintenanceConfiguration | Update-AzMaintenanceConfiguration -ResourceGroupName $resourceGroupName -Name $maintenanceConfigurationName, that's the case of reported issue.

File: src/Maintenance/Maintenance/ChangeLog.md [hunk 3]
```diff
@@ -20,6 +20,8 @@
 
 ## Upcoming Release
 * Fixed bug where rebootSettings property wasn't updating.
+* Fixed bug where Get-MaintenanceConfig was returning a List.
```
Comment (BethanyZhou) [inline src/Maintenance/Maintenance/ChangeLog.md:23]: Please always insert new change logs between L21-L22 to ensure new change logs is under the top of upcoming release section after merging into main branch

File: src/Maintenance/Maintenance/ChangeLog.md [hunk 4]
```diff
@@ -20,6 +20,8 @@
...
+    - This bug was affecting Update-MaintenanceConfig's input params.
```
Comment (BethanyZhou) [inline src/Maintenance/Maintenance/ChangeLog.md:24]:
```suggestion
    - This bug was affecting Update-MaintenanceConfig's input params.
```
This bug is not only affecting Update-MaintenanceConfiguration, but also affecting all cmdlets receive the return objects from Get-AzMaintenanceConfiguration like Remove-AzMaintenanceConfiguration

File: src/Maintenance/Maintenance/ChangeLog.md (hunk 3 above)
Comment (BethanyZhou) [inline src/Maintenance/Maintenance/ChangeLog.md:23]:
```suggestion
* Fixed bug where `Get-AzMaintenanceConfiguration` was returning a List. [#25781]
```
- ensure official name of cmdlet used in our documentation.
- no need to explain bug in change log but add the link of github issue.

**Review by BethanyZhou — CHANGES_REQUESTED** [review]

File: src/Maintenance/Maintenance/ChangeLog.md (hunk 3 above)
Comment (notyashhh) [reply src/Maintenance/Maintenance/ChangeLog.md:23]: Updated this in the next commit.

Comment (notyashhh) [reply src/Maintenance/Maintenance/ChangeLog.md:23]: Added this change :)

File: src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1 (hunk 2 above)
Comment (notyashhh) [reply src/Maintenance/Maintenance.Test/ScenarioTests/MaintenanceTests.ps1:85]: Updated this in the next commit. Thanks!

**Comment by BethanyZhou** [comment]
I think maybe we need two configurations in the test cases to check if the list will be unpacking as single enumeration

**Comment by BethanyZhou** [comment]
@notyashhh, if your PR is ready for reviewing again, add needs-review label.

**Review by BethanyZhou — DISMISSED** [review]
Good job. Looks good to me.

**Review by dolauli — APPROVED** [review]

**Pull request merged**